### DISCUSSION TITLE
meow keypad

### DISCUSSION THREAD

**Discussion by mclearc** [meow-edit/meow#107] [Ideas]
Hi -- thanks for meow. But I'm wondering whether meow keypad state is necessary. I'm thinking of cases where people use `<SPC>` as a leader or local leader key (this includes people who otherwise use emacs keybindings). I'm wondering if it would be possible to either (1) option (e.g. defcustom) to change the keypad leader key from `<SPC>` to something else; (2) option to disable keypad mode altogether. What do you think?

**Comment by DogLooksGood** [comment]
Hi,

The `SPC` is leader key, which is bound to `meow-leader-keymap`.

The entry point for keypad state is `meow-keypad-start`, which is bound to x, c, h, g, m in `meow-leader-keymap`. They are just normal keybindings. If you do 
```
(dolist (k '("x" "c" "h" "m" "g"))
  (define-key meow-leader-keymap k nil))
```
There will be no way to enter keypad state, thus you can consider it is disabled.

One of the main reasons behind meow is that I'm tired of writing hundreds of lines of configuration for just keybindings. I was trying to build a reasonable, beautiful, complete leader system when I was using Evil, but that is very time consuming, and not fun at all. So an idea came out, why not just using the existing keybindings. That's why there's a keypad state.

Meow is pretty opinionated, the only thing it has no opinion with, is the commands layout. That should be tweaking by users with their preference. But everything else is strongly opinioned, even though most are configurable. So meow doesn't have a major-mode specific leader keymap or local leader keymap and nested keymaps under leader is also not recommended.

I had to admire that some migration, especially from the Evil with complex leader system would be difficult. But if you look at [my emacs configuration](https://github.com/DogLooksGood/dogEmacs/blob/master/elisp/init-modal.el).
- All modal editing related configuration can be done in a single file. By excluding this file, you got clean, vanilla emacs.
- No boring keybindings everywhere

**Reply by mclearc** [reply]
Thanks very much for that tip! I certainly understand that one might not want to maintain keybindings, but I also think there is a lot to be gained by having commands grouped by use or application. You can see my evil keybindings here: https://github.com/mclear-tools/dotemacs/blob/master/setup-config/setup-keybindings.el. I'm trying to decide how much of this to keep. Trying out meow is letting me rethink a bunch of things but I suspect I'll keep at least some of these under a custom keybind (general makes changing the leader very easy).

**Reply by mclearc** [reply]
I'm getting a better understanding of `meow-leader-define-key` from looking at your config file -- thanks for sharing that!

**Reply by DogLooksGood** [reply]
You can still reuse your keybinding. For example, bind your custom org-mode keybindings to `C-c C-l` in `org-mode-map`, then bind `l` to `C-c C-l` in `meow-leader-keymap`. Then `l` will start keypad with `C-c C-l`. For example, `SPC l f` => `org-fill-paragraph`.

You group your keybindings by `C-c C-l`, `C-c C-c`, `C-c C-t`, etc. 

---

A lot packages already group their commands somewhere, you can bind by the entire keymap. You can bind a single key in leader to

- a dispatch command that shipped with package, e.g. `v`  to `C-x M-g`(magit-dispatch).
- an existing package keymap, e.g. `p` to `projectile-command-map`
- a nested sub keymap, e.g. `k` to `C-x k` (kmacro-keymap)

**Reply by mclearc** [reply]
Yep -- moved all my personal bindings behind `C-c C-SPC`, which is unused by most (all?) packages, and aliased that to `meow leader y`. Thanks again for your help.

**Reply by kotatsuyaki** [reply]
>     * a dispatch command that shipped with package, e.g. `v`  to `C-x M-g`(magit-dispatch).

I was looking for a way to run `magit-discard` in motion state using the default QWERTY layout without configuring anything further, and this gave me the hint - `? k` works out of the box, because `?` is bound to `magit-dispatch`. Thanks!
